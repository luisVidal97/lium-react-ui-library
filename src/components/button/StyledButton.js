import styled from 'styled-components';

export const StyledButton = styled.button`
  height: ${ props => props.shapeSize};
  font-size: ${ props => props.fontSize};
  color: ${ props => props.overallColor};
  background-color: ${ props => props.backgroundBtn};
  border: 2px solid ${ props => props.borderBtn};


  &:hover {
    transition: 0.4s;
    cursor: ${ props => props.cursor};
    background-color: ${ props => props.backgroundBtnHover};
    border: 2px solid ${ props => props.borderBtnHover};
    color: ${ props => props.cursor ==='pointer'? 'white' : props.overallColor };
  }

  &:focus { outline: none; }

  &:active {
    opacity: ${ props => props.cursor ==='pointer'? '0.4' : '1' };
    transition: 0.2s
  }

  border-radius:${ props => props.borderRadiusBtn};
  margin: 0 1em;
  padding: 0.25em 1em;
`;
