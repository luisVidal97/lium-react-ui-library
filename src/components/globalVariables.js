
const forms = ['contained', 'outlined'];

const heights = {
  large: '50px',
  medium: '40px',
  small: '30px',
}

const borderRadius = {
  strong: '20px',
  normal: '10px',
  none: '0px'
}

const heightsInputText = {
  large: {
    shape: '40px',
    font: '1.2rem'
  },
  medium:  {
    shape: '32px',
    font: '1rem'
  },
  small: {
    shape: '24px',
    font: '.8rem'
  },
}


const colors = {
  primary: {
      normal: '#115293',
      hover: '#0A0A36',
  },
  secondary: {
    normal: '#DC004E',
    hover: '#9A0036',
  },
  danger: {
      normal: '#ff0000',
      hover: '#cc0000',
  },
  warning: {
    normal: '#ffea05',
    hover: '#e2d006'
  },
  success: {
    normal: '#06c920',
    hover: '#019e16'
  },
  dark: {
    normal: '#000000',
    hover: '#686767',
  },
};

const defaultStyle = {
  normal: '#000000',
  hover: '#686767',
}

export { forms, borderRadius, heights, heightsInputText, colors, defaultStyle } ;
