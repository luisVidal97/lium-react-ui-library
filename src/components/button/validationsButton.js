import {colors, forms, heights} from '../globalVariables'


export const validations = ( {text, disabled: disabledBtn, shape, color, size} ) =>{

  text = text || '';
  shape = shape?
    forms.find(item => item === shape) || 'contained'
    : 'contained';

  color = color?
    colors[color] || colors['primary']
    : colors['primary'];

  let overallColor, backgroundBtn, backgroundBtnHover, borderBtn, borderBtnHover ;
  let cursor = 'pointer'
  if ( disabledBtn === 'true' &&  shape === 'contained') {
    overallColor = '#A6A6A6';
    backgroundBtn = '#E0E0E0';
    backgroundBtnHover =  '#E0E0E0';
    borderBtn =  '#E0E0E0';
    borderBtnHover =  '#E0E0E0';
    cursor = 'default';
  } else if(disabledBtn === 'true' &&  shape === 'outlined') {
    overallColor = '#E0E0E0';
    backgroundBtn ='white';
    backgroundBtnHover = 'white';
    borderBtn ='#E0E0E0';
    borderBtnHover = '#E0E0E0';
    cursor = 'default';
  } else if (shape === 'contained'){
    overallColor = 'white';
    backgroundBtn = color.normal;
    backgroundBtnHover = color.hover;
    borderBtn = color.normal;
    borderBtnHover = color.hover;
  } else{
    overallColor = color.normal;
    backgroundBtn ='white';
    backgroundBtnHover = color.hover;
    borderBtn = color.normal;
    borderBtnHover = color.hover;
  }

  size = size?
    heights[size] || heights['medium']
    : heights['medium'];

  return {
    text,
    shape,
    size ,
    overallColor,
    backgroundBtn,
    backgroundBtnHover,
    borderBtn,
    borderBtnHover,
    cursor
  };
}
