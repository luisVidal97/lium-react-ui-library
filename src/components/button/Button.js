import React, {Fragment} from 'react';
import { StyledButton } from './StyledButton';
import { validations } from './validationsButton';
import PropTypes from 'prop-types';

const Button = (props) => {

  const source = validations( props );

  return (
    <Fragment>
    {
      props.onClick && props.disabled !=='true'?
        <StyledButton {...source} onClick={props.onClick}>
          { source.text }
        </StyledButton>
      :
        <StyledButton {...source}>
          { source.text }
        </StyledButton>
    }
    </Fragment>
  )
}

Button.propTypes = {
  size: PropTypes.string,
  disabled: PropTypes.string,
  shape: PropTypes.string,
  text: PropTypes.string,
  color: PropTypes.string,
  onClick: PropTypes.func,
  border: PropTypes.string,
}

export default Button;
