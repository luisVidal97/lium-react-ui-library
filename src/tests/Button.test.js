import React from 'react';
import { shallow } from'enzyme';

import 'jest-styled-components'
import renderer from 'react-test-renderer'

import { heightComponents, colors } from '../components/globalVariables';
import  Button from '../components/button/Button';
import  {StyledButton} from '../components/button/StyledButton';


describe('Tests over <Button/>: features and behavior', () => {

  let wrapper = shallow(<Button />);
  let tree = renderer.create(<Button />).toJSON();

  test('should show <Button/> successfully', () => {
    wrapper = shallow(<Button />);
    expect( wrapper ).toMatchSnapshot();
  });


  test('should show default style values of button invoking <Button /> component (without props)', () => {
    tree = renderer.create(<Button />).toJSON();
    expect(tree).toHaveStyleRule('color', 'white');
    expect(tree).toHaveStyleRule('height', heightComponents.medium.shape);
    expect(tree).toHaveStyleRule('background-color', colors.primary.normal);
    expect(tree).toHaveStyleRule('background-color', colors.primary.hover, { modifier: ':hover', });
    expect(tree).toHaveStyleRule('cursor', 'pointer', { modifier: ':hover', });
  });

  test('should show default style values of button invoking <Button /> component (with props but wrong values)', () => {
    tree = renderer.create(<Button disabled='hi' size='sm' shape='cover' text='24' moove="large"/>).toJSON();
    expect(tree).toHaveStyleRule('color', 'white');
    expect(tree).toHaveStyleRule('height', heightComponents.medium.shape);
    expect(tree).toHaveStyleRule('background-color', colors.primary.normal);
    expect(tree).toHaveStyleRule('background-color', colors.primary.hover, { modifier: ':hover', });
    expect(tree).toHaveStyleRule('cursor', 'pointer', { modifier: ':hover', });
  });

  test('should show outlined, large and disabled button', () => {
    tree = renderer.create(<Button disabled='true' size='large' shape='outlined' />).toJSON();
    expect(tree).toHaveStyleRule('color', '#E0E0E0');
    expect(tree).toHaveStyleRule('height', heightComponents.large.shape);
    expect(tree).toHaveStyleRule('background-color', 'white');
    expect(tree).toHaveStyleRule('background-color',  'white', { modifier: ':hover', });
    expect(tree).toHaveStyleRule('cursor', 'default', { modifier: ':hover', });
  });

  test('should show "text prop" properly in the button, which it is send through props', () => {
    const helloText = 'Hello world!';
    wrapper = shallow( <Button text={helloText}/> );
    let textBtn = wrapper.find(StyledButton).text();
    expect(textBtn).toBe(helloText);

    wrapper = shallow( <Button /> );
    textBtn = wrapper.find(StyledButton).text();
    expect(textBtn).toBe('');
  });



  test('should execute the function when click the button', () => {

    const doSomething = jest.fn();
    expect( doSomething ).not.toHaveBeenCalled();

    wrapper = shallow( <Button onClick={ doSomething }/> );
    wrapper.find(StyledButton).simulate('click');
    expect( doSomething ).toHaveBeenCalled();

    wrapper.find(StyledButton).simulate('click');
    wrapper.find(StyledButton).simulate('click');
    expect( doSomething ).toHaveBeenCalledTimes(3)

  });


  test('should not execute the function when click the button if it is disabled', () => {

    const doSomething = jest.fn();

    wrapper = shallow( <Button onClick={ doSomething } disabled='true'/> );
    wrapper.find(StyledButton).simulate('click');
    expect( doSomething ).not.toHaveBeenCalled();

  });


});

