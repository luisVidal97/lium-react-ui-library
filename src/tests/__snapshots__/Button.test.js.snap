// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <Button/>: features and behavior should show <Button/> successfully 1`] = `
<Fragment>
  <styled.button
    backgroundBtn="#115293"
    backgroundBtnHover="#0A0A36"
    borderBtn="#115293"
    borderBtnHover="#0A0A36"
    cursor="pointer"
    overallColor="white"
    shape="contained"
    size="40px"
    text=""
  />
</Fragment>
`;
